File: packages/server/src/tmux/monitor.ts
import { createHash } from 'crypto';

export interface StatusUpdateResult {
  updated: boolean;
  hasPrompt: boolean;
}

/**
 * Monitors tmux session output for changes and prompt detection
 * Uses hash-based change detection for efficiency
 * Ported from go/pkg/tmux/monitor.go
 */
export class StatusMonitor {
  private prevOutputHash: string | null = null;

  /**
   * Check if content has changed and if there's a prompt waiting
   * @param content - Terminal output content to check
   * @returns Object with updated (content changed) and hasPrompt flags
   */
  hasUpdated(content: string): StatusUpdateResult {
    // Detect generic prompt patterns
    const trimmed = content.trim();
    const hasPrompt =
      trimmed.endsWith('>') || trimmed.endsWith('$') || trimmed.endsWith(':');

    // Check if content has changed using hash comparison
    const currentHash = this.hash(content);
    if (this.prevOutputHash !== currentHash) {
      this.prevOutputHash = currentHash;
      return { updated: true, hasPrompt };
    }

    return { updated: false, hasPrompt };
  }

  /**
   * Generate SHA-256 hash of content for efficient comparison
   */
  private hash(content: string): string {
    return createHash('sha256').update(content).digest('hex');
  }
}

File: go/pkg/tmux/monitor.go
// Package tmux provides tmux session management and monitoring functionality
// for creating, attaching to, and controlling tmux sessions.
package tmux

import (
	"crypto/sha256"
	"strings"
)

// StatusMonitor tracks changes in tmux session output
type StatusMonitor struct {
	prevOutputHash []byte
	program        string
}

// newStatusMonitor creates a new status monitor for a tmux session
func newStatusMonitor(program string) *StatusMonitor {
	return &StatusMonitor{
		prevOutputHash: make([]byte, 0),
		program:        program,
	}
}

// hash generates a SHA256 hash of the given content
func (m *StatusMonitor) hash(content string) []byte {
	h := sha256.Sum256([]byte(content))
	return h[:]
}

// HasUpdated checks if the content has changed and if there's a prompt waiting
func (m *StatusMonitor) HasUpdated(content string) (updated bool, hasPrompt bool) {
	// Check for prompts specific to different AI programs
	switch m.program {
	case "claude":
		hasPrompt = strings.Contains(content, "No, and tell Claude what to do differently")
	case "aider":
		hasPrompt = strings.Contains(content, "(Y)es/(N)o/(D)on't ask again")
	case "codex":
		// Add specific prompt detection for codex if needed
		hasPrompt = false
	default:
		// Generic prompt detection - look for common patterns
		hasPrompt = strings.HasSuffix(strings.TrimSpace(content), ">") ||
			strings.HasSuffix(strings.TrimSpace(content), "$") ||
			strings.HasSuffix(strings.TrimSpace(content), ":")
	}

	// Check if content has changed
	currentHash := m.hash(content)
	if !bytesEqual(currentHash, m.prevOutputHash) {
		m.prevOutputHash = currentHash
		return true, hasPrompt
	}

	return false, hasPrompt
}

// bytesEqual compares two byte slices for equality
func bytesEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
